ClosedXML/Excel/CalcEngine/Functions/Tally.cs

```diff
@@ -56,85 +61,87 @@ public void Add(Expression e)
             else
                 foreach (var v in valEnumerable)
                     _list.Add(v);
+
+            _numericValues = null;
         }
 
         public void AddValue(Object v)
         {
             _list.Add(v);
+            _numericValues = null;
         }
 
         public double Count()
         {
-            return this.Count(this.NumbersOnly);
+            return Count(NumbersOnly);
         }
 
         public double Count(bool numbersOnly)
         {
             if (numbersOnly)
-                return NumericValues().Count();
+                return NumericValues().Count;
             else
-                return _list.Where(o => !Statistical.IsBlank(o)).Count();
+                return _list.Count(o => !Statistical.IsBlank(o));
         }
 
-        public IEnumerable<Double> NumericValues()
+        IEnumerable<double> NumericValuesInternal()
         {
-            var retVal = new List<double>();
             foreach (var value in _list)
             {
-                Double tmp;
                 var vEnumerable = value as IEnumerable;
-                if (vEnumerable == null && Double.TryParse(value.ToString(), out tmp))
-                    yield return tmp;
+                if (vEnumerable == null)
+                {
+                    if (double.TryParse(value.ToString(), out double tmp))
+                        yield return tmp;
+                }
                 else
                 {
                     foreach (var v in vEnumerable)
                     {
-                        if (Double.TryParse(v.ToString(), out tmp))
+                        if (double.TryParse(v.ToString(), out double tmp))
                             yield return tmp;
                         break;
                     }
                 }
             }
         }
 
+        public IReadOnlyList<double> NumericValues()
```

Comment:
IReadOnlyList<T> was introduced in .NET 4.5, it does not seem to compile with .NET 4.0.

Reply:
Ok. I'll make an update.